Don't print progress on non-interactive terminals

MichaelEischer opened the pull request:
What does this PR change? What problem does it solve?
-----------------------------------------------------
The PR revert the progress bar updates to the old behavior of not printing progress updates on non-interactive terminals.

As an additional change, the `RESTIC_PROGRESS_FPS` environment variable is now also interpreted for non-interactive terminals. This makes it possible to use the environment variable to also get regular progress updates on non-interactive terminals.

Was the change discussed in an issue or in the forum before?
------------------------------------------------------------
Fixes #3194.
See #2706 for further discussions.

Checklist
---------

- [x] I have read the [Contribution Guidelines](https://github.com/restic/restic/blob/master/CONTRIBUTING.md#providing-patches)
- [x] I have enabled [maintainer edits for this PR](https://help.github.com/en/github/collaborating-with-issues-and-pull-requests/allowing-changes-to-a-pull-request-branch-created-from-a-fork)
- [x] I have added tests for all changes in this PR **We currently have no means for testing the `progress.go` change, however, I've verified that it works manually.**
- [x] I have added documentation for the changes (in the manual)
- [x] There's a new file in `changelog/unreleased/` that describes the changes for our users (template [here](https://github.com/restic/restic/blob/master/changelog/TEMPLATE))
- [x] I have run `gofmt` on the code in all commits
- [x] All commit messages are formatted in the same style as [the other commits in the repo](https://github.com/restic/restic/blob/master/CONTRIBUTING.md#git-commits)
- [x] I'm done, this Pull Request is ready for review

MichaelEischer committed:
ui/progress: don't print progress on non-interactive terminals

This reverts to the old behavior of not printing progress updates on
non-interactive terminals. It was accidentally changed in #3058.

MichaelEischer committed:
Honor RESTIC_PROGRESS_FPS env variable on non-interactive terminals

This makes it possible to use the environment variable to also get
regular progress updates on non-interactive terminals.

fd0 commented:
Code looks good already!

DRON-666 commented:
If I understand correctly, this PR is primary intended for Unix users, but it would be nice to add Windows support.
Now on Windows output redirected to a pipe always contains progress lines and redirected to a file never contains it (this PR doesn't change anything in this behavior). I think it's better to disable progress on redirected output by default and enable it via `RESTIC_PROGRESS_FPS`.

MichaelEischer committed:
Unify progress report frequency calculation

MichaelEischer committed:
ui: cleanup backup status shutdown

MichaelEischer committed:
ui: update status for the backup command on non-interactive terminals

Allow the backup command to print status on non-interactive terminals.
The output is disabled by setting a MinUpdatePause == 0.

MichaelEischer committed:
ui: don't shorten non-interactive progress output

MichaelEischer committed:
ui/progress: extract signal handling into own package

MichaelEischer committed:
ui: force backup progress update on signal

MichaelEischer committed:
Support values less than 1 for RESTIC_PROGRESS_FPS

For example set the variable to 0.016666 to print the progress once per
minute.

MichaelEischer committed:
ui: remove dead struct member

MichaelEischer committed:
Add changelog entry

MichaelEischer commented:
@DRON-666 That behavior on Windows is currently intentional. Please have a look at https://github.com/restic/restic/blob/a488d4c847ed9631bab537d0e3b9ec654bed978a/internal/ui/termstatus/terminal_windows.go#L97 . The problem is that the detection of whether a terminal is interactive or not is not always reliable. However, that problem is out of scope for this PR.

I've updated the PR to also adapt the progress output of the `backup` command which uses its own implementation. The progress report frequency is now also controlled in a similar manner to the generic progress bar implementation.

`RESTIC_PROGRESS_FPS` now uses floating-point numbers to allow reducing the progress report frequency to less than once per second.

DRON-666 commented:
>However, that problem is out of scope for this PR.

You right. I will write a new PR after merging of this.

fd0 committed:
ui: Simlify channel receive

fd0 reviewed (approved):
Great work, thanks! I've added a tiny improvement on top, but then it's good to go!
